src/generic/locked_fs.ts

```diff
@@ -0,0 +1,398 @@
+import Mutex from './mutex';
+import {FileSystem, SynchronousFileSystem} from '../core/file_system';
+import {ApiError, ErrorCode} from '../core/api_error';
+import {FileFlag, ActionType} from '../core/file_flag';
+import {default as Stats, FileType} from '../core/node_fs_stats';
+import {File} from '../core/file';
+
+// TODO: also implement LockedFile to ensure that operations on files
+// also grab the appropriate locks?.
+/*
+export LockedFile<T extends File> implements File {
+  private _file: File;
+  private _mu: Mutex;
+
+  constructor(mu: Mutex, f: File) {
+    this._file = f;
+    this._mu = mu;
+  }
+
+  getPos(): number;
+
+  stat(cb: (err: ApiError, stats?: Stats) => any): void;
+
+  statSync(): Stats;
+
+  close(cb: Function): void;
+
+  closeSync(): void;
+
+  truncate(len: number, cb: Function): void;
+
+  truncateSync(len: number): void;
+
+  sync(cb: (e?: ApiError) => void): void;
+
+  syncSync(): void;
+
+  write(buffer: NodeBuffer, offset: number, length: number, position: number, cb: (err: ApiError, written?: number, buffer?: NodeBuffer) => any): void;
+
+  writeSync(buffer: NodeBuffer, offset: number, length: number, position: number): number;
+
+  read(buffer: NodeBuffer, offset: number, length: number, position: number, cb: (err: ApiError, bytesRead?: number, buffer?: NodeBuffer) => void): void;
+
+  readSync(buffer: NodeBuffer, offset: number, length: number, position: number): number;
+
+  datasync(cb: (e?: ApiError) => void): void;
+
+  datasyncSync(): void;
+
+  chown(uid: number, gid: number, cb: (e?: ApiError) => void): void;
+
+  chownSync(uid: number, gid: number): void;
+
+  chmod(mode: number, cb: (e?: ApiError) => void): void;
+
+  chmodSync(mode: number): void;
+
+  utimes(atime: Date, mtime: Date, cb: (e?: ApiError) => void): void;
+
+  utimesSync(atime: Date, mtime: Date): void;
+}
+*/
+
+/// This class serializes access to an underlying async filesystem.
+/// For example, on an OverlayFS instance with an async lower
+/// directory operations like rename and rmdir may involve multiple
+/// requests involving both the upper and lower filesystems -- they
+/// are not executed in a single atomic step.  OverlayFS uses this
+/// LockedFS to avoid having to reason about the correctness of
+/// multiple requests interleaving.
+export class LockedFS<T extends FileSystem> implements FileSystem {
+  private _fs: T;
+  private _mu: Mutex;
+
+  constructor(cls: {new(...args: any[]): T}, ...args: any[]) {
+    this._fs = new (Function.prototype.bind.apply(cls, [null].concat(args)));
```

Comment:
`Function.prototype.bind` is not supported in IE8. Please add a polyfill to our polyfill file.